src/Compilers/CSharp/Portable/Compilation/SyntaxTreeSemanticModel.cs

```diff
@@ -1759,6 +1759,12 @@ private Symbol GetDeclaredMember(NamespaceOrTypeSymbol container, TextSpan decla
 
             // look for any member with same declaration location
             var collection = name != null ? container.GetMembers(name) : container.GetMembersUnordered();
+            if (isKnownNamespace && name != null && collection is [NamespaceSymbol knownNamespace])
```

Comment:
📝 In the heap dump I collected at a sample point during a very long build (almost 3 hours), all 8 threads sitting in this member would have hit this special case and returned the correct value immediately.